### PR TITLE
Feat/add pkh suport for deploy composite 

### PR DESCRIPTION
**Pull request by qbig**
# Add did:pkh Support for create composite models - #[Issue]

## Description

https://forum.ceramic.network/t/only-admin-could-create-models-composites/1135/22

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [ ] Test A (e.g. Test A - New test that ... ran in local, docker, and dev unstable.)
- [ ] Test B

## Definition of Done

Before submitting this PR, please make sure:

- [ ] The work addresses the description and outcomes in the issue
- [ ] I have added relevant tests for new or updated functionality
- [ ] My code follows conventions, is well commented, and easy to understand
- [ ] My code builds and tests pass without any errors or warnings
- [ ] I have tagged the relevant reviewers
- [ ] I have updated the READMEs of affected packages
- [ ] I have made corresponding changes to the documentation
- [ ] The changes have been communicated to interested parties

## References:

Please list relevant documentation (e.g. tech specs, articles, related work etc.) relevant to this change, and note if the documentation has been updated.

### CODE REVIEW

**Review by oed — CHANGES_REQUESTED** [review]
We should not need a network call for this

File: packages/devtools/src/composite.ts
Comment (oed) [reply packages/devtools/src/composite.ts:100]: @PaulLeCam Looks like we got some type signatures incorrect?

Comment (qbig) [reply packages/devtools/src/composite.ts:100]: [SignedCommitContainer](https://github.com/ceramicnetwork/js-ceramic/blob/03b544717b295eb5134a890720c4b15e278f6ca1/packages/common/src/stream.ts#LL63C1-L63C1) was cast to SignedCommit via here `commit as unknown as SignedCommit`
```typescript
      const [genesis, ...updates] = commits as Array<StreamCommits>
      const model = await ceramic.createStreamFromGenesis<Model>(
        Model.STREAM_TYPE_ID,
        genesis,
        MODEL_GENESIS_OPTS
      )
      for (const commit of updates) {
        const signedCommit = commit as unknown as SignedCommit
        assertSupportedReadModelController(model, signedCommit)
        await ceramic.applyCommit(model.id, signedCommit)
      }
```

Comment (PaulLeCam) [reply packages/devtools/src/composite.ts:100]: Yes definitely possible these types are incorrect, don't hesitate to fix in this PR

File: packages/devtools/src/composite.ts
Comment (PaulLeCam) [inline packages/devtools/src/composite.ts:26]:
```suggestion
const MODEL_STREAM_ID = Model.MODEL.toString()
```
Better use the stream ID defined on the model itself than copying it I think.

File: packages/devtools/src/composite.ts
Comment (PaulLeCam) [inline packages/devtools/src/composite.ts:88]:
```suggestion
    if (!hasModelResource) {
```
Seems incorrect to ensure only one resource is allowed in the CACAO, no? The CACAO could be used for other resources as well.

File: packages/devtools/src/composite.ts
Comment (PaulLeCam) [inline packages/devtools/src/composite.ts:95]:
```suggestion
      `${unsupported}, only did:key and did:pkh are supported`
```

**Review by PaulLeCam — CHANGES_REQUESTED** [review]
Thanks for your PR, looks good overall but I wonder if this logic should live on the Model handler implementation rather than here?
It would be safer to ensure all models are validated rather than only the ones read or written using our devtools, what do you think @oed please?

**Comment by oed** [comment]
> I wonder if this logic should live on the Model handler implementation rather than here?

@PaulLeCam I think this would only be a minor improvement since any streamtype specific logic won't be enforced on the protocol level after the migration to the rust implementation. I think we could add a check there, but I think logic here will be more impactful.

File: packages/devtools/src/composite.ts
Comment (oed) [reply packages/devtools/src/composite.ts:100]: @qbig If you console.log the commit I think you will get a [SignedCommitContainer](https://github.com/ceramicnetwork/js-ceramic/blob/03b544717b295eb5134a890720c4b15e278f6ca1/packages/common/src/stream.ts#LL63C1-L63C1)?

Comment (oed) [reply packages/devtools/src/composite.ts:100]: Yep, SignedCommitContainer = DagJWSResult.

So you can decode the `cacaoBlock` (which will be present when did:pkh is used) using dag-cbor to get the Cacao object.

File: packages/devtools/src/composite.ts
Comment (oed) [inline packages/devtools/src/composite.ts:87]: I think this logic is backwards? We should only allow creation if ocap doesn't have expiry?
```suggestion
    if (cacao.p.exp !== null) {
```

File: packages/devtools/src/composite.ts
Comment (oed) [inline packages/devtools/src/composite.ts:134]:
```suggestion
      `only cacao with resource "ceramic://*?model=${MODEL_STREAM_ID}" is supported`
```
We can make this error more explicit!

File: packages/devtools/src/composite.ts
Comment (oed) [inline packages/devtools/src/composite.ts:126]:
```suggestion
  if (cacao.p.exp !== null) {
```
This one also looks incorrect?

File: packages/devtools/src/composite.ts
Comment (oed) [inline packages/devtools/src/composite.ts:186]: Shouldn't we only do this check if it's a signed container? AnchorCommits won't be signed.

Comment (oed) [reply packages/devtools/src/composite.ts:186]: We also need to check if `genesis` with `assertSupportedReadModelController`.

**Review by oed — CHANGES_REQUESTED** [review]
This looks much better. A few updates needed.

File: packages/devtools/src/composite.ts
Comment (qbig) [reply packages/devtools/src/composite.ts:186]: > Shouldn't we only do this check if it's a signed container? AnchorCommits won't be signed.

Do we need to validate AnchorCommit here? possible to do it without network request?

Comment (qbig) [reply packages/devtools/src/composite.ts:186]: > Shouldn't we only do this check if it's a signed container? AnchorCommits won't be signed.
I was assuming they are all signed commit from the function signature. How do I handle the anchor commit here ?
```typescript
async function loadModelsFromCommits<Models = Record<string, StreamCommits>>(
  ceramic: CeramicApi,
  modelsCommits: Models & {} // eslint-disable-line @typescript-eslint/ban-types
): Promise<Record<keyof Models, ModelDefinition>> {
//...
export type StreamCommits = Array<DagJWSResult>;
//...
export declare type SignedCommitContainer = DagJWSResult;

```

Comment (qbig) [reply packages/devtools/src/composite.ts:186]: > We also need to check if `genesis` with `assertSupportedReadModelController`.

How about `static async fromModels(params: FromModelsParams): Promise<Composite> ` Do we also need to validate AnchorCommit and genesis commit? Does `loadStreamCommits` only return genesis and signed commits?

Comment (PaulLeCam) [reply packages/devtools/src/composite.ts:188]: `loadStreamCommits()` loads all the commits, including anchor ones, but they are filters to only signed commits in the following logic.
Assuming there is no expiration date on the CACAO I don't think we need to care about the anchor commits, unless I'm missing something? cc @oed

Comment (oed) [reply packages/devtools/src/composite.ts:188]: We don't need to validate the anchor commits.

**Review by oed — DISMISSED** [review]
lgtm, would be good to get another review from @PaulLeCam as well.

File: packages/devtools/src/composite.ts
Comment (PaulLeCam) [inline packages/devtools/src/composite.ts:126]:
```suggestion
  if (cacao.p.exp != null) {
```

**Review by PaulLeCam — CHANGES_REQUESTED** [review]
Thanks for the changes, mostly LGTM but I just spotted a few things.

**Review by oed — DISMISSED** [review]
I think this should be good to go. Will let @PaulLeCam have another look before we merge.

**Review by PaulLeCam — APPROVED** [review]
Thanks for the changes!

**Comment by PaulLeCam** [comment]
@qbig looks like there is an issue with the lockfile, can you make sure it is updated please?

**Comment by PaulLeCam** [comment]
@qbig could you make sure your branch is in sync with the `main` branch and passes the checks locally (running `pnpm install`, `pnpm run lint` and `pnpm test`) please?

**Review by PaulLeCam — APPROVED** [review]

**Comment by PaulLeCam** [comment]
Thanks @qbig!

**Pull request merged**
